src/skadi/sample.cc

```diff
@@ -72,42 +68,105 @@ namespace {
   }
 
   uint64_t file_size(const std::string& file_name) {
+    //TODO: detect gzip and actually validate the uncompressed size?
     struct stat s;
     int rc = stat(file_name.c_str(), &s);
     return rc == 0 ? s.st_size : -1;
   }
 
+  void gunzip(const valhalla::midgard::mem_map<char>& in, int16_t* tile) {
+    //make a zstream with in and out
+    z_stream stream = {
+      static_cast<Byte*>(static_cast<void*>(in.get())),
+        static_cast<unsigned int>(in.size()), static_cast<unsigned int>(in.size()), //the input stream
+      static_cast<Byte*>(static_cast<void*>(tile)), HGT_BYTES, HGT_BYTES //the output stream
+    };
+
+    //decompress the file
+    if(inflateInit2(&stream, 16 + MAX_WBITS) != Z_OK)
+      throw std::runtime_error("gzip decompression init failed");
+    auto e = inflate(&stream, Z_FINISH);
+    if(e != Z_STREAM_END || stream.total_out != HGT_BYTES)
+      throw std::runtime_error("Corrupt gzip elevation data");
+    inflateEnd(&stream);
+  }
+
+  void lunzip(const valhalla::midgard::mem_map<char>& in, int16_t* tile) {
+    auto compressed_size = LZ4_decompress_fast(in.get(), static_cast<char*>(static_cast<void*>(tile)), HGT_BYTES);
+    if(compressed_size != in.size())
+      throw std::runtime_error("Corrupt lz4 elevation data");
+  }
+
 }
 
 namespace valhalla {
 namespace skadi {
 
-  sample::sample(const std::string& data_source):cache(TILE_COUNT) {
+  sample::sample(const std::string& data_source):
+      mapped_cache(TILE_COUNT), unzipped(-1, std::vector<int16_t>(HGT_PIXELS)) {
     //check the directory for files that look like what we need
     auto files = get_files(data_source);
     for(const auto& f : files) {
-      auto index = is_hgt(f);
-      if(index < cache.size()) {
-        if(file_size(f) == HGT_PIXELS * sizeof(int16_t))
-          cache[index].map(f, HGT_PIXELS * sizeof(int16_t));
-        else
+      //make sure its a valid index
+      format_t format;
+      auto index = is_hgt(f, format);
+      if(index < mapped_cache.size() && format > mapped_cache[index].first) {
+        auto size = file_size(f);
+        if(format == format_t::RAW && size != HGT_BYTES) {
           LOG_WARN("Corrupt elevation data: " + f);
+          continue;
+        }
+        mapped_cache[index].first = format;
+        mapped_cache[index].second.map(f, size, POSIX_FADV_SEQUENTIAL);
```

Comment:
when mapping we tell the kernel to make the read ahead double its size. this makes it so that we are very likely to have an entire `lz4` compressed tile in ram